### PR TITLE
performance improvements for FeatureFinderMultiplex

### PR DESCRIPTION
**Pull request by lars20070**
PR #3644 fixed a bug by simply removing a [pre-filtering step](https://github.com/OpenMS/OpenMS/compare/develop...lars20070:fix/FFM_performance?expand=1#diff-d07005aa96a93e57990ba6d481c8425eR66). (The peak map is pre-filtered using an intensity cutoff. This step reduces memory footprint and runtime.)

This PR re-introduces the pre-filtering step and fixes the old bug by passing the filtered peaks and peak boundaries to the subsequent `featureXML` and `consensusXML` map construction.

Results remain the same, see unchanged `TOPP` tests.

### CODE REVIEW

File: src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.h [hunk 1]
```diff
@@ -103,13 +103,17 @@ namespace OpenMS
...
      * @param averagine_similarity_scaling    scaling factor x for the averagine similarity parameter p when detecting peptide singlets. With p' = p + x(1-p). 
      */
-    MultiplexFiltering(const MSExperiment& exp_picked, const std::vector<MultiplexIsotopicPeakPattern>& patterns, int isotopes_per_peptide_min,
+    MultiplexFiltering(const MSExperiment& exp_centroided, const std::vector<MultiplexIsotopicPeakPattern>& patterns, int isotopes_per_peptide_min,
                        int isotopes_per_peptide_max, double intensity_cutoff, double rt_band, double mz_tolerance, bool mz_tolerance_unit,
                        double peptide_similarity, double averagine_similarity, double averagine_similarity_scaling, String averagine_type="peptide");
+    /**
+     * @brief returns the intensity-filtered, centroided spectral data
+     */
+    MSExperiment getCentroidedExperiment();
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.h]: don't return a copy here but a const reference

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderMultiplexAlgorithm.cpp [hunk 2]
```diff
@@ -755,54 +755,53 @@ namespace OpenMS
...
+       */
       MultiplexClustering clustering(exp_centroid_, param_.getValue("algorithm:mz_tolerance"), (param_.getValue("algorithm:mz_unit") == "ppm"), param_.getValue("algorithm:rt_typical"), static_cast<double>(param_.getValue("algorithm:rt_min")));
       clustering.setLogType(getLogType());
-      cluster_results = clustering.cluster(filter_results);
+      std::vector<std::map<int, GridBasedCluster> > cluster_results = clustering.cluster(filter_results);
+
+      /**
+       * construct feature and consensus maps i.e. the final results
+       */
+      filtering.getCentroidedExperiment().swap(exp_centroid_);
```
Comment (timosachsenberg) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderMultiplexAlgorithm.cpp]: what should this line do? swapping exp_centroid_ with the copy?

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderMultiplexAlgorithm.cpp [hunk 3]
```diff
@@ -755,54 +755,53 @@ namespace OpenMS
...
-    else
-    {
-      //consensus_map.setPrimaryMSRunPath(exp_profile_.getPrimaryMSRunPath());
-      //feature_map.setPrimaryMSRunPath(exp_profile_.getPrimaryMSRunPath());
+      std::vector<std::map<int, GridBasedCluster> > cluster_results = clustering.cluster(filter_results);
+      
+      /**
+       * construct feature and consensus maps i.e. the final results
+       */
+      filtering.getCentroidedExperiment().swap(exp_centroid_);
```
Comment (timosachsenberg) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderMultiplexAlgorithm.cpp]: same here... a temporary copy is swapped in? does this make sense?

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp [hunk 4]
```diff
@@ -51,19 +51,42 @@ using namespace boost::math;
...
-  exp_picked_(exp_picked), patterns_(patterns), isotopes_per_peptide_min_(isotopes_per_peptide_min), isotopes_per_peptide_max_(isotopes_per_peptide_max),
+  patterns_(patterns), isotopes_per_peptide_min_(isotopes_per_peptide_min), isotopes_per_peptide_max_(isotopes_per_peptide_max),
   intensity_cutoff_(intensity_cutoff), rt_band_(rt_band), mz_tolerance_(mz_tolerance), mz_tolerance_unit_in_ppm_(mz_tolerance_unit),
   peptide_similarity_(peptide_similarity), averagine_similarity_(averagine_similarity),
   averagine_similarity_scaling_(averagine_similarity_scaling), averagine_type_(averagine_type)
   {
+    // initialise experiment exp_centroided_
+    // Any peaks below the intensity cutoff cannot be relevant. They are therefore removed resulting in reduced memory footprint and runtime.
+    // loop over spectra
+    for (MSExperiment::ConstIterator it_rt = exp_centroided.begin(); it_rt < exp_centroided.end(); ++it_rt)
```
Comment (timosachsenberg) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp]: can you use our ThresholdMower algorithm in OpenMS

Comment (timosachsenberg) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp]: usually != x.end() is used and not < x.end()

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp [hunk 5]
```diff
@@ -51,19 +51,42 @@ using namespace boost::math;
...
+      spectrum.setRT(it_rt->getRT());
+      // loop over m/z
+      for (MSSpectrum::ConstIterator it_mz = it_rt->begin(); it_mz < it_rt->end(); ++it_mz)
+      {
+        if (it_mz->getIntensity() > intensity_cutoff_)
+        {
+          Peak1D peak;
+          peak.setMZ(it_mz->getMZ());
+          peak.setIntensity(it_mz->getIntensity());
+          spectrum.push_back(peak);
```
Comment (timosachsenberg) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp]: why not push_back(*it_mz)?

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp [hunk 6]
```diff
@@ -76,27 +99,32 @@ namespace OpenMS
...
+          blacklist_[it_rt - exp_centroided_.begin()][it_mz - it_rt->begin()] = 666;
         }
       }
     }
     
   }
+  
+  MSExperiment MultiplexFiltering::getCentroidedExperiment()
+  {
+    return(exp_centroided_);
```
Comment (timosachsenberg) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp]: we usually don't use () around return value

Comment (timosachsenberg) [reply src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp]: consider returning a const reference here

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp [hunk 7]
```diff
@@ -76,27 +99,32 @@ namespace OpenMS
...
+  }
 
   void MultiplexFiltering::updateWhiteMSExperiment_()
   {
     // reset both the white MS experiment and the corresponding mapping to the complete i.e. original MS experiment
-    exp_picked_white_.clear(true);
-    exp_picked_mapping_.clear();
+    exp_centroided_white_.clear(true);
```
Comment (timosachsenberg) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp]: comment why in one case true is passed

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringProfile.cpp [hunk 8]
```diff
@@ -243,6 +279,11 @@ namespace OpenMS
 
     return filter_results;
   }
+  
+  std::vector<std::vector<PeakPickerHiRes::PeakBoundary> > MultiplexFilteringProfile::getPeakBoundaries()
```
Comment (timosachsenberg) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringProfile.cpp]: return a const reference?

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp [hunk 9]
```diff
@@ -51,19 +51,42 @@ using namespace boost::math;
...
                                          int isotopes_per_peptide_min, int isotopes_per_peptide_max, double intensity_cutoff, double rt_band,
                                          double mz_tolerance, bool mz_tolerance_unit, double peptide_similarity, double averagine_similarity,
                                          double averagine_similarity_scaling, String averagine_type) :
-  exp_picked_(exp_picked), patterns_(patterns), isotopes_per_peptide_min_(isotopes_per_peptide_min), isotopes_per_peptide_max_(isotopes_per_peptide_max),
+  patterns_(patterns), isotopes_per_peptide_min_(isotopes_per_peptide_min), isotopes_per_peptide_max_(isotopes_per_peptide_max),
   intensity_cutoff_(intensity_cutoff), rt_band_(rt_band), mz_tolerance_(mz_tolerance), mz_tolerance_unit_in_ppm_(mz_tolerance_unit),
   peptide_similarity_(peptide_similarity), averagine_similarity_(averagine_similarity),
   averagine_similarity_scaling_(averagine_similarity_scaling), averagine_type_(averagine_type)
   {
+    // initialise experiment exp_centroided_
```
Comment (hroest) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp]: to save memory and speed up calculations, I suggest here:

```
exp_centroided_.reserve(exp_centroided.getNrSpectra())
```

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp [hunk 10]
```diff
@@ -51,19 +51,42 @@ using namespace boost::math;
...
+      {
+        if (it_mz->getIntensity() > intensity_cutoff_)
+        {
+          Peak1D peak;
+          peak.setMZ(it_mz->getMZ());
+          peak.setIntensity(it_mz->getIntensity());
+          spectrum.push_back(peak);
+        }
+      }
+      exp_centroided_.addSpectrum(spectrum);
```
Comment (hroest) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp]: I suggest to change this to

```
exp_centroided_.addSpectrum(std::move(spectrum));
```

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp (hunk 6 above)
Comment (hroest) [reply src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp]: > consider returning a const reference here

yes, unless you want to make a copy I also suggest a const ref

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringProfile.cpp [hunk 11]
```diff
@@ -40,34 +40,70 @@
...
+    // (We run a simple intensity filter. Peaks below the intensity cutoff can be discarded right from the start.)
+    // Now we still need to discard boundaries of low intensity peaks, in order to preserve the one-to-one mapping between peaks and boundaries.
     
-    if (exp_profile.size() != exp_picked.size())
+    // loop over spectra and boundaries
+    MSExperiment::ConstIterator it_rt;
+    std::vector<std::vector<PeakPickerHiRes::PeakBoundary> >::const_iterator it_rt_boundaries;
+    for (it_rt = exp_centroided.begin(), it_rt_boundaries = boundaries.begin();
+         it_rt < exp_centroided.end(), it_rt_boundaries < boundaries.end();
+         ++it_rt, ++it_rt_boundaries)
```
Comment (hroest) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringProfile.cpp]: I suggest

```
for (auto it_rt = exp_centroided.begin(), auto it_rt_boundaries = boundaries.begin();
         it_rt != exp_centroided.end() && it_rt_boundaries != boundaries.end();
        ++it_rt, ++it_rt_boundaries)
```

I dont think you want the comma operator here:

```
In a comma expression E1, E2, the expression E1 is evaluated, its result is discarded
```

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringProfile.cpp [hunk 12]
```diff
@@ -40,34 +40,70 @@
...
+    {
+      // new boundaries of a single spectrum
+      std::vector<PeakPickerHiRes::PeakBoundary> boundaries_temp;
+      
+      // loop over m/z peaks and boundaries
+      MSSpectrum::ConstIterator it_mz;
+      std::vector<PeakPickerHiRes::PeakBoundary>::const_iterator it_mz_boundaries;
+      for (it_mz = it_rt->begin(), it_mz_boundaries = it_rt_boundaries->begin();
+           it_mz < it_rt->end(), it_mz_boundaries < it_rt_boundaries->end();
+           ++it_mz, ++it_mz_boundaries)
```
Comment (hroest) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringProfile.cpp]: same here, use `auto`, use `!=` and dont use comma

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringProfile.cpp [hunk 13]
```diff
@@ -40,34 +40,70 @@
...
+
+          // Check consistency of peaks and their peak boundaries, i.e. check that the peak lies in the boundary interval.
+          if (((*it_mz_boundaries).mz_min > it_mz->getMZ()) || (it_mz->getMZ() > (*it_mz_boundaries).mz_max))
+          {
+            throw Exception::InvalidRange(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION);
+          }
+        }
+      }
+      
+      boundaries_.push_back(boundaries_temp);
```
Comment (hroest) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringProfile.cpp]: can you `reserve` this before the loop?

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderMultiplexAlgorithm.cpp (hunk 2 above)
Comment (hroest) [reply src/openms/source/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderMultiplexAlgorithm.cpp]: is this what fixes the bug?

Comment (lars20070) [reply src/openms/source/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderMultiplexAlgorithm.cpp]: `exp_centroid_` is updated by the (intensity filtered) centroid data from the `filtering` object.

Comment (lars20070) [reply src/openms/source/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderMultiplexAlgorithm.cpp]: @hroest Yes, the filter results are stored as [indices](https://github.com/OpenMS/OpenMS/blob/develop/src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/MultiplexSatelliteCentroided.h#L61) in the **filtered** centroid data. Hence, I update this member variable before constructing the maps.

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringProfile.cpp (hunk 11 above)
Comment (lars20070) [reply src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringProfile.cpp]: Cannot get the double auto loop to work. Complains about second auto.
```
/Users/lars/Code/OpenMS-develop/OpenMS/src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringProfile.cpp:62:47: error: expected
      unqualified-id
    for (auto it_rt = exp_centroided.begin(), auto it_rt_boundaries = boundaries.begin();
```

Comment (timosachsenberg) [reply src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringProfile.cpp]: please don't use the comma operator in for loops. And it's not about the auto...

for (int i = 0, int b = 0; i != 10; ++i)
{
}
will also not compile.

Comment (lars20070) [reply src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringProfile.cpp]: Ok. I will loop over one container and access the other two via indices.

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp (hunk 7 above)
Comment (lars20070) [reply src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp]: The first one is an `MSExperiment`, the second one is a `vector`.
`true` clears the metadata as well.

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp (hunk 9 above)
Comment (lars20070) [reply src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFiltering.cpp]: [Done.](https://github.com/OpenMS/OpenMS/pull/3768/files#diff-d07005aa96a93e57990ba6d481c8425eR65)

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringProfile.cpp (hunk 13 above)
Comment (lars20070) [reply src/openms/source/TRANSFORMATIONS/FEATUREFINDER/MultiplexFilteringProfile.cpp]: [Done/](https://github.com/OpenMS/OpenMS/pull/3768/files#diff-d3f64d534dd7f30a56e4fac9c0d7a5c8R58)

**Comment by lars20070** [comment]
@timosachsenberg @hroest Thanks for the good comments.
Hope the changes make the code more readable and faster. Sorry for the micro-commits.

**Comment by timosachsenberg** [comment]
+1 if tests pass

**Comment by jpfeuffer** [comment]
rebuild jenkins

**Comment by lars20070** [comment]
Two remaining build errors. All tests pass.

**Pull request merged**